server/athenian/api/experiments/preloading/cache.py

```diff
@@ -0,0 +1,259 @@
+import asyncio
+from typing import Dict, List, Optional, Union
+
+import pandas as pd
+import sqlalchemy as sa
+from sqlalchemy import select
+
+from athenian.api.typing_utils import DatabaseLike
+
+
+def _humanize_size(v):
+    unit = ""
+    for u in ["B", "KB", "MB", "GB"]:
+        unit = u
+        if v < 1024:
+            break
+        v = v / 1024
+
+    return f"{round(v, 2)} {unit}"
+
+
+class CachedDataFrame:
+    """A DataFrame that is loaded and can be refreshed from the database.
+
+    This can be read as a normal `pandas.DataFrame`.
+
+    """
+
+    def __init__(
+        self,
+        id_: str,
+        db: DatabaseLike,
+        cols: Optional[List[sa.Column]] = None,
+        timestamp_cols: Optional[List[sa.Column]] = None,
+        categorical_cols: Optional[List[sa.Column]] = None,
+        identifier_cols: Optional[List[sa.Column]] = None,
+        filtering_clause: Optional[sa.sql.ClauseElement] = None,
+        debug_memory: Optional[bool] = False,
+    ):
+        """Initialize a `CachedDataFrame`."""
+        self._id = id_
+        self._cols = cols or ["*"]
+        self._timestamp_cols = timestamp_cols or [
+            c for c in cols if isinstance(c.property.columns[0].type, sa.TIMESTAMP)
+        ]
+        self._categorical_cols = categorical_cols or []
+        self._identifier_cols = identifier_cols or []
+        self._filtering_clause = filtering_clause
+        self._db = db
+        self._debug_memory = debug_memory
+        self._mem = {}
+        self._df = None
+
+    @property
+    def id_(self) -> str:
+        """Return the id of the `CachedDataFrame`."""
+        return self._id
+
+    async def load(self) -> "CachedDataFrame":
+        """Load the DataFrame from the database."""
+        query = select(self._cols)
+        if self._filtering_clause:
+            query = query.where(self._filtering_clause)
+
+        res = await self._db.fetch_all(query)
+        df = pd.DataFrame(res)
+        if self._debug_memory:
+            size = df.memory_usage(index=True, deep=True)
+            self._mem["raw"] = {"series": size, "total": sum(size)}
+
+        self._df = self._squeeze(df)
+        if self._debug_memory:
+            size = self._df.memory_usage(index=True, deep=True)
+            self._mem["processed"] = {"series": size, "total": sum(size)}
```

Comment:
There should be a way to write this to our `/status` using Prometheus. See `status_controller.py`

Reply:
Make sense, but I'd rather do this separately. I'll create a separate ticket for this.

Reply:
https://athenianco.atlassian.net/browse/DEV-2167